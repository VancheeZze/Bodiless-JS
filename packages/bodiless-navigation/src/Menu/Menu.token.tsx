/**
 * Copyright Â© 2020 Johnson & Johnson
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { omit } from 'lodash';
import { useEditContext } from '@bodiless/core';
import {
  addClasses,
  removeClassesIf,
  addClassesIf,
  withDesign,
  TokenDef,
  asToken,
  Design,
} from '@bodiless/fclasses';

import { useIsMenuOpen } from './withMenuContext';
import {
  asFlex, asOverflowHidden, asRelative, asAbsolute,
  asPositionedLeft, withFullWidthStyles, withColumnDirectionStyles,
  withStaticOnHoverStyles, withVisibleOnHoverStyles, asElementToken, asListToken,
} from '../token';

/**
 * Helper which makes it easier to target a particular type of submenu.
 *
 * The first parameter is a list of the submenu key(s) to which the token
 * provided as a second argument should be applied.
 * It also accepts the special key 'Main' to apply the design to the top level menu.
 *
 * @param keys List of the submenu key(s) to which the token should be applied.
 * @param tokenDefs List of tokens to be applied to submenu design key(s).
 *
 * @return Desigh token that applies supplied list of tokens to the provided design keys.
 */
export const withSubMenuToken = <P extends object>(
  ...keys: string[]
) => (...tokenDefs: TokenDef<P>[]) => {
    const design: Design<any> = keys.reduce((acc, key) => ({
      ...acc,
      [key]: asToken(...tokenDefs),
    }), {});

    const withSubMenuDesign = withDesign({
      Item: withDesign(omit(design, 'Main')),
    });

    /* eslint-disable dot-notation */
    return design['Main']
      ? asToken(design['Main'], withSubMenuDesign)
      : withSubMenuDesign;
  };

/*
 * Utility Styles
 * ===========================================
 */
const isContextActive = () => {
  const { isActive, isEdit } = useEditContext();
  return isEdit && isActive;
};

const isContextNotActive = () => {
  const { isActive, isEdit } = useEditContext();
  return isEdit ? !isActive : true;
};

const asVerticalSubMenu = withDesign({
  Wrapper: withColumnDirectionStyles,
});

const asVisibleOnActive = asToken(
  addClassesIf(isContextActive)('overflow-visible'),
  asElementToken('Layout')('Overflow'),
);

const asResponsiveSublist = withDesign({
  Wrapper: asToken(
    addClasses('min-w-full'),
    asElementToken('Sizing')('Min-Width'),
  ),
});

const asStaticOnHover = asToken(
  withStaticOnHoverStyles,
  removeClassesIf(useIsMenuOpen)('hover:static'),
  asElementToken('Layout')('Position'),
);

const asRelativeNotActive = asToken(
  addClassesIf(isContextNotActive)('relative'),
  asElementToken('Layout')('Position'),
);

const asFullWidthSublist = withDesign({
  Wrapper: withFullWidthStyles,
});

const withHoverStyles = withDesign({
  Item: asToken(
    withVisibleOnHoverStyles,
    removeClassesIf(useIsMenuOpen)('hover:overflow-visible'),
    asElementToken('Layout')('Overflow'),
  ),
});

/*
 * Base Menu Styles
 * ===========================================
 */
const withBaseMenuStyles = asToken(
  withHoverStyles,
  withDesign({
    Wrapper: asToken(asFlex, asRelative),
    Item: asOverflowHidden,
  }),
  // @todo Discuss Attr and Category
  asListToken('Menu')('Main'),
);

/*
 * Base Sub Menu Styles
 * ===========================================
 */
const withBaseSubMenuStyles = withDesign({
  Wrapper: asToken(asFlex, asAbsolute, asPositionedLeft),
});

/*
 * List Sub Menu Styles
 * ===========================================
 */
const asListSubMenu = asToken(
  asResponsiveSublist,
  asVerticalSubMenu,
  withBaseSubMenuStyles,
  asVisibleOnActive,
  asRelative,
  asListToken('Submenu')('List'),
);

/*
 * Touts Sub Menu Styles
 * ===========================================
 */
const asToutsSubMenu = asToken(
  asFullWidthSublist,
  asStaticOnHover,
  withBaseSubMenuStyles,
  asRelativeNotActive,
  asListToken('Submenu')('Touts'),
);

/*
 * Columns Sub Menu Styles
 * ===========================================
 */
const asColumnSubMenu = asToken(
  asFullWidthSublist,
  asStaticOnHover,
  withBaseSubMenuStyles,
  asRelativeNotActive,
  asListToken('Submenu')('Columns'),
);

/**
 * Helper which allows specifying which submenu types should have default navigation styling added.
 *
 * @param keys List of the submenu key(s) to which the default menu styles be applied to.
 *
 * @return Token that applies default top navigation styles based on provided keys.
 */
export const asTopNav = (...keys: string[]) => {
  const listSubmenuStyles = keys.indexOf('List') > -1 ? asListSubMenu : asToken();
  const toutsSubmenuStyles = keys.indexOf('Touts') > -1 ? asToutsSubMenu : asToken();
  const columnsSubmenuStyles = keys.indexOf('Columns') > -1 ? asColumnSubMenu : asToken();

  return asToken(
    withSubMenuToken('Main')(withBaseMenuStyles),
    withSubMenuToken('List')(listSubmenuStyles),
    withSubMenuToken('Touts')(toutsSubmenuStyles),
    withSubMenuToken('Columns')(columnsSubmenuStyles),
  );
};
