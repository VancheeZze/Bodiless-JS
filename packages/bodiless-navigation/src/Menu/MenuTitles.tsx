/**
 * Copyright Â© 2020 Johnson & Johnson
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { ComponentType, FC } from 'react';
import { flow } from 'lodash';
import {
  withSidecarNodes, withNode, withNodeKey, WithNodeProps,
} from '@bodiless/core';
import { asEditable, asBodilessLink, withBodilessLinkToggle } from '@bodiless/components';
import { ToutClean } from '@bodiless/organisms';
import {
  A,
  Div,
  HOC,
  Design,
  asToken,
  Fragment,
  designable,
  withDesign,
  startWith,
  DesignableProps,
  DesignableComponentsProps,
} from '@bodiless/fclasses';

import { useBreadcrumbContext, asBreadcrumbSource as asBreadcrumbSourceBase } from '../Breadcrumbs';

type MenuTitleComponents = {
  Link: ComponentType<any>,
  Title: ComponentType<any>,
};

type MenuTitleProps = DesignableComponentsProps<MenuTitleComponents>;

const DEFAULT_NODE_KEYS = {
  toggleNodeKey: 'title$link-toggle',
  linkNodeKey: 'title$link',
  titleNodeKey: 'title$text',
};

/**
 * Hook which can be used to determine if a menu item is part of
 * the current active breadcrumb trail.
 *
 * This hook is only accurate if
 * - The menu is inside a BreadcrumbStoreProvider.
 * - The menu item has been wrapped in asBreadcrumb
 *
 * @return true if the item is in the active trail, false otherwise.
 */
const useIsActiveTrail = () => useBreadcrumbContext()?.isActive();

const asBreadcrumbSource = flow(
  asBreadcrumbSourceBase(DEFAULT_NODE_KEYS),
  withNode,
);

const MenuTitleBase: FC<MenuTitleProps> = ({ components, ...rest }) => {
  const { Link, Title } = components;
  return (
    <Link {...rest}>
      <Title />
    </Link>
  );
};

const MenuTitleComponents: MenuTitleComponents = {
  Link: A,
  Title: Fragment,
};

const MenuTitle = designable(MenuTitleComponents, 'MenuTitle')(MenuTitleBase);

const asMenuLink = (asEditableLink: typeof asBodilessLink) => asToken(
  withSidecarNodes(
    asEditableLink('link', undefined, () => ({ groupLabel: 'Menu Link' })),
  ),
);

const asMenuTitle = (asEditableTitle: typeof asEditable) => asToken(
  asEditableTitle('text', 'Menu Item'),
);

const asEditableMenuTitle = flow(
  startWith(MenuTitle),
  withDesign({
    Link: asMenuLink(withBodilessLinkToggle(asBodilessLink, startWith(Div) as HOC)),
    Title: asMenuTitle(asEditable),
  }),
  withNode,
  withNodeKey('title'),
);

const withEditableMenuTitle = withDesign({
  Title: asEditableMenuTitle,
});

const asMenuTout = (linkNodeKey = 'link', titleNodeKey = 'text') => {
  const transformDesign = (design: Design<any> = {}) => {
    const Link = asToken(
      withSidecarNodes(
        design.Link || withBodilessLinkToggle(asBodilessLink, startWith(Div) as HOC)(),
        withNodeKey(linkNodeKey),
      ),
    );
    const Title = asToken(
      design.Title || asEditable(undefined, 'Menu Item'),
      withNodeKey(titleNodeKey),
    );
    return { ...design, Link, Title };
  };

  const CleanTout: ComponentType<DesignableProps<any> & WithNodeProps> = ({ design, ...rest }) => (
    <ToutClean design={transformDesign(design)} {...rest} />
  );

  return startWith(CleanTout);
};

export default MenuTitle;
export {
  DEFAULT_NODE_KEYS,
  asMenuLink,
  asMenuTitle,
  asMenuTout,
  useIsActiveTrail,
  asEditableMenuTitle,
  withEditableMenuTitle,
  asBreadcrumbSource,
};
