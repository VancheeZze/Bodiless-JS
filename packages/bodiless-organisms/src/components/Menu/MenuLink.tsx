/**
 * Copyright Â© 2019 Johnson & Johnson
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { FC, ComponentType, HTMLProps } from 'react';
import {
  A,
  designable,
  DesignableComponentsProps,
  addClasses,
} from '@bodiless/fclasses';
import { useBreadcrumbStore, useBreadcrumbContext } from '@bodiless/components';
import { observer } from 'mobx-react-lite';
import { flow } from 'lodash';

type MenuLinkComponents = {
  Link: ComponentType<any>,
  ActiveLink: ComponentType<any>,
  ActiveTrailLink: ComponentType<any>,
};

const menulinkComponents: MenuLinkComponents = {
  Link: A,
  ActiveLink: A,
  ActiveTrailLink: addClasses('text-orange-700')(A),
};

const isCurrentPage = (href: string | undefined) => {
  if (!href || (typeof window === 'undefined')) {
    return false;
  }
  const urls = [];
  urls.push(window.location.toString());

  let relativeUrl = window.location.pathname;
  urls.push(relativeUrl);
  if (window.location.search) {
    relativeUrl += window.location.search;
    urls.push(relativeUrl);
  }
  if (window.location.hash) {
    relativeUrl += window.location.hash;
    urls.push(relativeUrl);
    urls.push(window.location.hash); // a case when only a hash presents in link href
  }
  return urls.indexOf(href) > -1;
};

export type Props = {
  href?: string,
  unwrap?: Function,
} & DesignableComponentsProps<MenuLinkComponents> & HTMLProps<HTMLElement>;

/**
 * Toggles between two states. Render <ActiveLink /> when the active page URL (window.location)
 * matches to href attribute of the component. Otherwise the <Link /> will be rendered
 * Note - below the href prop is passed in by the withData() HOC which is part of asBodilessLink.
 * @constructor
 */
const MenuLinkBase: FC<Props> = ({ components, unwrap, ...rest }) => {
  const { Link, ActiveLink, ActiveTrailLink } = components;
  const { href } = rest;
  const item = useBreadcrumbContext();
  const store = useBreadcrumbStore();

  if (isCurrentPage(href)) {
    return <ActiveLink {...rest} />;
  }
  if (item && store && store.breadcrumbTrail.find(tItem => tItem.isEqual(item))) {
    return <ActiveTrailLink {...rest} />;
  }
  return <Link {...rest} />;
};

const MenuLink = flow(
  observer,
  designable(menulinkComponents),
)(MenuLinkBase);

export default MenuLink;
