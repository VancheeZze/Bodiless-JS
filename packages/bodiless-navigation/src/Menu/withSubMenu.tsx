/**
 * Copyright Â© 2020 Johnson & Johnson
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { flow } from 'lodash';
import {
  HOC, Design, withDesign, asToken,
} from '@bodiless/fclasses';
import {
  asStylableList, asStylableSubList, asSubList, withDeleteNodeOnUnwrap,
  withSubLists, UseListOverrides,
} from '@bodiless/components';

import { asBreadcrumbSource, withEditableMenuTitle } from './MenuTitles';

/**
 * Creates a stylable sublist which deletes it's data when the last item is removed.
 * Suitable for use for all menus.
 *
 * @param titleDesign optional token which will be applied to the sublist title.
 * @param useOverrides optional hook returning overrides for the submenu button.
 */
const asMenuSubList = (titleDesign: any = {}, useOverrides: UseListOverrides = () => ({})) => {
  const titleDesign$ = typeof titleDesign === 'function' ? titleDesign : withDesign(titleDesign);
  return flow(
    asSubList((props) => ({ groupLabel: 'Sub-Menu Item', ...useOverrides(props) })),
    asStylableList,
    asStylableSubList,
    withDeleteNodeOnUnwrap('sublist'),
    withEditableMenuTitle,
    asBreadcrumbSource,
    withDesign({
      Title: titleDesign$,
    }),
  );
};

const withSubMenuDesign = (design: Design<any>) => withDesign({
  Item: withDesign(design),
});

/**
 * Helper which can be used to add a List submenu option to the menu.
 *
 * @param menuTitleDesign? optional token which will be applied to the sublist title.
 *
 */
const withListSubMenu = (menuTitleDesign?: Design<any> | HOC) => withSubMenuDesign({
  List: asMenuSubList(menuTitleDesign),
});

/**
 * Helper which can be used to add a Touts submenu option to the menu.
 *
 * @param menuTitleDesign? optional token which will be applied to the sublist title.
 *
 */
const withToutSubMenu = (menuTitleDesign?: Design<any> | HOC) => withSubMenuDesign({
  Touts: asMenuSubList(menuTitleDesign),
});

/**
 * Helper which can be used to add a Columns submenu option to the menu.
 *
 * @param menuTitleDesign? optional token which will be applied to the sublist title.
 *
 */
const withColumnSubMenu = (menuTitleDesign?: Design<any> | HOC) => withSubMenuDesign({
  Columns: asToken(
    asMenuSubList(menuTitleDesign),
    withSubLists(1)(asMenuSubList(menuTitleDesign, () => ({ groupLabel: 'Column Sub-Menu Item' }))),
  ),
});

export {
  withListSubMenu,
  withToutSubMenu,
  withColumnSubMenu,
  withSubMenuDesign,
};
