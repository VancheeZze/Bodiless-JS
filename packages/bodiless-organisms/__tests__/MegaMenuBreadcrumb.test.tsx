/**
 * Copyright Â© 2020 Johnson & Johnson
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { mount } from 'enzyme';
import { withDefaultContent, withSidecarNodes } from '@bodiless/core';
import { asBodilessLink } from '@bodiless/components';
import { replaceWith, withDesign, withOnlyProps } from '@bodiless/fclasses';
import { flowRight } from 'lodash';

import { asBreadcrumbsClean, asMenuBase } from '../src/components/Menu/MegaMenu';

const { DefaultContentNode } = require('@bodiless/core');

const setPagePath = (pagePath: string) => {
  Object.defineProperty(DefaultContentNode.prototype, 'pagePath', {
    value: pagePath,
    writable: false,
  });
};

const createBreadcrumbComponent = ({
  content = {},
}) => flowRight(
  withDefaultContent(content),
  withDesign({
    // @ts-ignore
    BreadcrumbLink: replaceWith(withSidecarNodes(
      asBodilessLink(),
    ))('a'),
    BreadcrumbTitle: replaceWith(withOnlyProps('key', 'children')(React.Fragment)),
  }),
  asBreadcrumbsClean({
    linkNodeKey: 'title$link',
    titleNodeKey: 'title$text',
  }),
  asMenuBase('testMenu'),
)('ul');

describe('asBreadcrumbsClean', () => {
  it('creates breadcrumbs for 2-level Touts MegaMenu', () => {
    setPagePath('/products/productA');
    const Breadcrumb = createBreadcrumbComponent({
      content: {
        testMenu: {
          items: [
            'home',
            'products',
          ],
        },
        testMenu$home$title$link: {
          href: '/',
        },
        testMenu$products$title$link: {
          href: '/products',
        },
        testMenu$products$sublist: {
          items: [
            'productA',
            'productB',
          ],
        },
        'testMenu$products$cham-sublist': {
          component: 'Touts',
        },
        testMenu$products$sublist$productA$title$link: {
          href: '/products/productA',
        },
        testMenu$products$sublist$productB$title$link: {
          href: '/products/productB',
        },
      },
    });
    const wrapper = mount(<Breadcrumb />);
    expect(wrapper.html()).toMatchSnapshot();
  });
  it('creates breadcrumbs for 2-level List MegaMenu', () => {
    setPagePath('/products/productA');
    const Breadcrumb = createBreadcrumbComponent({
      content: {
        testMenu: {
          items: [
            'home',
            'products',
          ],
        },
        testMenu$home$title$link: {
          href: '/',
        },
        testMenu$products$title$link: {
          href: '/products',
        },
        testMenu$products$sublist: {
          items: [
            'productA',
            'productB',
          ],
        },
        'testMenu$products$cham-sublist': {
          component: 'List',
        },
        testMenu$products$sublist$productA$title$link: {
          href: '/products/productA',
        },
        testMenu$products$sublist$productB$title$link: {
          href: '/products/productB',
        },
      },
    });
    const wrapper = mount(<Breadcrumb />);
    expect(wrapper.html()).toMatchSnapshot();
  });
  it('creates breadcrumbs for 2-level Columns MegaMenu', () => {
    setPagePath('/products/productA');
    const Breadcrumb = createBreadcrumbComponent({
      content: {
        testMenu: {
          items: [
            'home',
            'products',
          ],
        },
        testMenu$home$title$link: {
          href: '/',
        },
        testMenu$products$title$link: {
          href: '/products',
        },
        testMenu$products$sublist: {
          items: [
            'productA',
            'productB',
          ],
        },
        'testMenu$products$cham-sublist': {
          component: 'Columns',
        },
        testMenu$products$sublist$productA$title$link: {
          href: '/products/productA',
        },
        testMenu$products$sublist$productB$title$link: {
          href: '/products/productB',
        },
      },
    });
    const wrapper = mount(<Breadcrumb />);
    expect(wrapper.html()).toMatchSnapshot();
  });
});
