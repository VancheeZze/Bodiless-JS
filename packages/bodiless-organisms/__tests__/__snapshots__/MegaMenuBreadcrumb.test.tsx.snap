// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`asBreadcrumbsClean creates breadcrumbs for basic 2-level menu 1`] = `"<ul><li><a href=\\"/products\\"></a></li><span></span><li><a href=\\"/products/productA\\"></a></li></ul>"`;
